File: API/Shipments/Services/ShipmentService.cs
using Transflower.EAgroServices.Shipments.Services.Interfaces;
using Transflower.EAgroServices.Shipments.Repositories.Interfaces;
using Transflower.EAgroServices.Shipments.Models;
using Transflower.EAgroServices.Shipments.Extensions;
using Transflower.EAgroServices.Shipments.Entities;

namespace Transflower.EAgroServices.Shipments.Services;

public class ShipmentService : IShipmentService
{
    private readonly IShipmentRepository _repository;

    public ShipmentService(IShipmentRepository repository)
    {
        _repository = repository;
    }

    public async Task<List<Shipment>> GetAll()
    {
        return await _repository.GetAll();
    }

    public async Task<Shipment?> GetById(int shipmentId)
    {
        return await _repository.GetById(shipmentId);
    }

    public async Task<List<InprogressShipment>> GetInprogressShipments()
    {
        return await _repository.GetInprogressShipments();
    }

    public async Task<PagedList<ShippedCollection>> GetShippedCollections(
        int collectionCenterId,
        string shipmentStatus,
        FilterRequest request,
        int pageNumber
    )
    {
        return await _repository.GetShippedCollections(
            collectionCenterId,
            shipmentStatus,
            request,
            pageNumber
        );
    }

    public async Task<List<ShipmentItemDetail>> GetShipmentItemsById(int shipmentId)
    {
        return await _repository.GetShipmentItemsById(shipmentId);
    }

    public async Task<TransporterAmount?> GetTransporterAmountByShipmentId(int shipmentId)
    {
        return await _repository.GetTransporterAmountByShipmentId(shipmentId);
    }

    public async Task<bool> IsShipmentStatusDelivered(int shipmentId)
    {
        return await _repository.IsShipmentStatusDelivered(shipmentId);
    }

    public async Task<bool> UpdateStatus(int shipmentId, UpdateStatus statusObject)
    {
        return await _repository.UpdateStatus(shipmentId, statusObject);
    }

    public async Task<bool> Insert(Shipment shipment)
    {
        return await _repository.Insert(shipment);
    }

    public async Task<bool> Update(Shipment shipment)
    {
        return await _repository.Update(shipment);
    }

    public async Task<bool> Delete(int shipmentId)
    {
        return await _repository.Delete(shipmentId);
    }

    public async Task<List<CorporateShipment>> GetShipmentByVehicleId(int vehicleId)
    {
        return await _repository.GetShipmentByVehicleId(vehicleId);
    }

    public async Task<List<VehicleCorporateShipment>> GetShipmentofTransporter(int transporterId)
    {
        return await _repository.GetShipmentofTransporter(transporterId);
    }

    public async Task<List<MerchantShipment>?> GetInprogressShipmentsByMerchant(int merchantId)
    {
        return await _repository.GetInprogressShipmentsByMerchant(merchantId);
    }

    public async Task<List<MerchantShipment>?> GetDeliveredShipmentsByMerchant(
        int merchantId,
        string paymentStatus
    )
    {
        return await _repository.GetDeliveredShipmentsByMerchant(merchantId, paymentStatus);
    }
}

File: UI/EAgroUIApp/src/app/Services/shipment.service.ts
import { HttpClient, HttpParams, HttpResponse } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable, forkJoin, map, of, switchMap } from 'rxjs';
import { MerchantShipment } from '../Models/merchant-shipment';
import { ShipmentItemDetails } from '../Models/shipment-item-details';
import { TransporterAmount } from '../Models/transporter-amount';
import { InprogressVehicle } from '../Models/inprogress-vehicle';
import { ShipmentItem } from '../Models/shipment-item';
import { FilterRequest } from '../Shared/filter/filter-request';
import { ShipmentStatus } from '../Models/Enums/shipment-status';
import { ShippedCollection } from '../Models/shipped-collection';
import { CorporateService } from './corporate.service';
import { UserService } from '../Shared/users/user.service';

@Injectable({
  providedIn: 'root',
})
export class ShipmentService {
  constructor(
    private http: HttpClient,
    private corporatesvc: CorporateService,
    private usersvc: UserService
  ) {}

  getInprogressShipmentsByMerchant(): Observable<MerchantShipment[]> {
    const merchantId = localStorage.getItem('merchantId');
    let url =
      'http://localhost:5067/api/shipments/inprogress/merchant/' + merchantId;
    return this.http.get<MerchantShipment[]>(url);
  }
  getDeliveredShipmentByMerchant(
    paymentStatus: string
  ): Observable<MerchantShipment[]> {
    const merchantId = localStorage.getItem('merchantId');
    let url =
      'http://localhost:5067/api/shipments/delivered/merchant/' +
      merchantId +
      '/' +
      paymentStatus;
    return this.http.get<MerchantShipment[]>(url);
  }

  getShipmentItems(shipmentId: number): Observable<ShipmentItemDetails[]> {
    let url = 'http://localhost:5067/api/shipments/shipmentitems/' + shipmentId;
  
    return this.http.get<ShipmentItemDetails[]>(url).pipe(
      switchMap((shipmentItemsDetails) => {
        if (shipmentItemsDetails.length === 0) {
          return of([]); 
        }
  
        const collectionCenterIds = new Set<number>();
        const farmerIds = new Set<number>();
  
        shipmentItemsDetails.forEach((item) => {
          collectionCenterIds.add(item.collectionCenterCorporaterId);
          farmerIds.add(item.farmerId);
        });
  
        const collectionCenterIdString = Array.from(collectionCenterIds).join(',');
        const farmerIdString = Array.from(farmerIds).join(',');
  
        const corporations$ = this.corporatesvc.getCorporates(collectionCenterIdString);
        const farmers$ = this.usersvc.getUserNamesWithId(farmerIdString);
  
        return forkJoin([corporations$, farmers$]).pipe(
          map(([corporationNames, farmerNames]) => {
            shipmentItemsDetails.forEach((item) => {
              const matchingCorporation = corporationNames.find(
                (element) => element.id === item.collectionCenterCorporaterId
              );
              if (matchingCorporation) {
                item.collectionCenterName = matchingCorporation.name;
              }
  
              const matchingFarmer = farmerNames.find(
                (element) => element.id === item.farmerId
              );
              if (matchingFarmer) {
                item.farmerName = matchingFarmer.name;
              }
            });
            return shipmentItemsDetails;
          })
        );
      })
    );
  }
  
  removeShipmentItem(shipmentItemId: number): Observable<boolean> {
    let url = 'http://localhost:5067/api/shipmentItems/' + shipmentItemId;
    return this.http.delete<boolean>(url);
  }

  isShipmentStatusDelivered(shipmentId: number): Observable<boolean> {
    let url = 'http://localhost:5067/api/shipments/status/' + shipmentId;
    return this.http.get<boolean>(url);
  }

  updateShipmentStatus(shipmentId: number): Observable<boolean> {
    let obj = {
      status: ShipmentStatus.delivered,
    };
    let url = 'http://localhost:5067/api/shipments/status/' + shipmentId;
    return this.http.patch<boolean>(url, obj);
  }

  getShipmentTransporterAmount(
    shipmentId: number
  ): Observable<TransporterAmount> {
    let url =
      'http://localhost:5067/api/shipments/transporteramount/' + shipmentId;
    return this.http.get<TransporterAmount>(url);
  }

  addShipment(shipment: any): Observable<boolean> {
    let url = 'http://localhost:5067/api/shipments';
    return this.http.post<boolean>(url, shipment);
  }

  getInprogressShipments(): Observable<InprogressVehicle[]> {
    let url = 'http://localhost:5067/api/shipments/inprogress';
    return this.http.get<any>(url);
  }

  addShipmentItem(shipmentItem: ShipmentItem): Observable<boolean> {
    let url = 'http://localhost:5067/api/shipmentitems';
    return this.http.post<boolean>(url, shipmentItem);
  }

  getShippedCollections(
    filterRequest: FilterRequest,
    pageNumber: number,
    staus: string
  ): Observable<HttpResponse<ShippedCollection[]>> {
    let collectionCenterId = localStorage.getItem('collectionCenterId');
    let url =
      'http://localhost:5067/api/shipments/collections/' +
      collectionCenterId +
      '/status/' +
      staus;
    let params = new HttpParams().set('pageNumber', pageNumber.toString());
    return this.http.post<ShippedCollection[]>(url, filterRequest, {
      params: params,
      observe: 'response',
    });
  }
}
